File: src/main/java/com/procurement/budget/service/EiServiceImpl.java
package com.procurement.budget.service;

import com.datastax.driver.core.utils.UUIDs;
import com.procurement.budget.config.properties.OCDSProperties;
import com.procurement.budget.dao.EiDao;
import com.procurement.budget.exception.ErrorException;
import com.procurement.budget.exception.ErrorType;
import com.procurement.budget.model.dto.bpe.ResponseDto;
import com.procurement.budget.model.dto.ei.EiDto;
import com.procurement.budget.model.dto.ei.EiOrganizationReferenceDto;
import com.procurement.budget.model.dto.ocds.TenderStatus;
import com.procurement.budget.model.dto.ocds.TenderStatusDetails;
import com.procurement.budget.model.entity.EiEntity;
import com.procurement.budget.utils.DateUtil;
import com.procurement.budget.utils.JsonUtil;
import java.time.LocalDateTime;
import java.util.Optional;
import java.util.UUID;
import org.springframework.stereotype.Service;

@Service
public class EiServiceImpl implements EiService {

    private static final String SEPARATOR = "-";
    private final OCDSProperties ocdsProperties;
    private final JsonUtil jsonUtil;
    private final DateUtil dateUtil;
    private final EiDao eiDao;

    public EiServiceImpl(final OCDSProperties ocdsProperties,
                         final JsonUtil jsonUtil,
                         final DateUtil dateUtil,
                         final EiDao eiDao) {
        this.ocdsProperties = ocdsProperties;
        this.jsonUtil = jsonUtil;
        this.dateUtil = dateUtil;
        this.eiDao = eiDao;
    }

    @Override
    public ResponseDto createEi(final String owner,
                                final String country,
                                final LocalDateTime dateTime,
                                final EiDto ei) {
        final String cpId = getCpId(country);
        ei.setOcId(cpId);
        setTenderId(ei, cpId);
        setTenderStatus(ei);
        setBudgetId(ei);
        setIdOfOrganizationReference(ei.getBuyer());
        validatePeriod(ei);
        final EiEntity entity = getEntity(ei, owner, dateTime);
        eiDao.save(entity);
        ei.setToken(entity.getToken().toString());
        return new ResponseDto<>(true, null, ei);
    }

    @Override
    public ResponseDto updateEi(final String owner,
                                final String cpId,
                                final String token,
                                final EiDto eiDto) {

        final EiEntity entity = Optional.ofNullable(eiDao.getByCpIdAndToken(cpId, UUID.fromString(token)))
                .orElseThrow(() -> new ErrorException(ErrorType.DATA_NOT_FOUND));
        if (!entity.getOwner().equals(owner)) throw new ErrorException(ErrorType.INVALID_OWNER);
        final EiDto ei = jsonUtil.toObject(EiDto.class, entity.getJsonData());
        ei.setPlanning(eiDto.getPlanning());
        ei.setTender(eiDto.getTender());
        entity.setJsonData(jsonUtil.toJson(ei));
        eiDao.save(entity);
        return new ResponseDto<>(true, null, ei);
    }

    @Override
    public EiDto getEi(final String cpId) {
        final EiEntity entity = Optional.ofNullable(eiDao.getByCpId(cpId))
                .orElseThrow(() -> new ErrorException(ErrorType.DATA_NOT_FOUND));
        return jsonUtil.toObject(EiDto.class, entity.getJsonData());
    }

    private String getCpId(final String country) {
        return ocdsProperties.getPrefix() + SEPARATOR + country + SEPARATOR +
                dateUtil.milliNowUTC();
    }

    private void setIdOfOrganizationReference(final EiOrganizationReferenceDto or) {
        or.setId(or.getIdentifier().getScheme() + SEPARATOR + or.getIdentifier().getId());
    }

    private void setTenderId(final EiDto ei, final String cpId) {
        ei.getTender().setId(cpId);
    }

    private void setTenderStatus(final EiDto ei) {
        ei.getTender().setStatus(TenderStatus.PLANNING);
        ei.getTender().setStatusDetails(TenderStatusDetails.EMPTY);
    }

    private void setBudgetId(final EiDto ei) {
        ei.getPlanning().getBudget().setId(ei.getTender().getClassification().getId());
    }

    private void validatePeriod(final EiDto ei) {
        if (!ei.getPlanning().getBudget().getPeriod().getStartDate()
                .isBefore(ei.getPlanning().getBudget().getPeriod().getEndDate()))
            throw new ErrorException(ErrorType.INVALID_PERIOD);
    }

    private EiEntity getEntity(final EiDto ei, final String owner, final LocalDateTime dateTime) {
        final EiEntity eiEntity = new EiEntity();
        eiEntity.setCpId(ei.getOcId());
        eiEntity.setToken(UUIDs.random());
        eiEntity.setOwner(owner);
        eiEntity.setCreatedDate(dateUtil.localToDate(dateTime));
        eiEntity.setJsonData(jsonUtil.toJson(ei));
        return eiEntity;
    }

}

File: src/main/kotlin/com/procurement/budget/service/EiServiceImpl.kt
package com.procurement.budget.service

import com.datastax.driver.core.utils.UUIDs
import com.procurement.budget.config.OCDSProperties
import com.procurement.budget.dao.EiDao
import com.procurement.budget.model.dto.ei.EiDto
import com.procurement.budget.model.dto.ocds.TenderStatus
import com.procurement.budget.model.dto.ocds.TenderStatusDetails
import com.procurement.budget.model.entity.EiEntity
import com.procurement.budget.utils.milliNowUTC
import com.procurement.budget.utils.toDate
import com.procurement.budget.utils.toJson
import com.procurement.budget.utils.toObject
import com.procurement.notice.exception.ErrorException
import com.procurement.notice.exception.ErrorType
import com.procurement.notice.model.bpe.ResponseDto
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*

interface EiService {

    fun createEi(owner: String,
                 country: String,
                 dateTime: LocalDateTime,
                 ei: EiDto): ResponseDto<*>

    fun updateEi(owner: String,
                 cpId: String,
                 token: String,
                 eiDto: EiDto): ResponseDto<*>

    fun getEi(cpId: String): EiDto
}

@Service
class EiServiceImpl(private val ocdsProperties: OCDSProperties,
                    private val eiDao: EiDao,
                    private val generateServise: GenerateServise) : EiService {

    override fun createEi(owner: String,
                          country: String,
                          dateTime: LocalDateTime,
                          ei: EiDto): ResponseDto<*> {
        validatePeriod(ei)
        val cpId = getCpId(country)
        ei.apply {
            ocid = cpId
            tender.id = cpId
            tender.status = TenderStatus.PLANNING
            tender.statusDetails = TenderStatusDetails.EMPTY
            planning.budget.id = tender.classification.id
            buyer.apply { id = identifier.scheme + SEPARATOR + identifier.id }
        }
        val entity = getEntity(ei, owner, dateTime)
        eiDao.save(entity)
        ei.token = entity.token.toString()
        return ResponseDto(true, null, ei)
    }

    override fun updateEi(owner: String,
                          cpId: String,
                          token: String,
                          eiDto: EiDto): ResponseDto<*> {
        val entity = eiDao.getByCpId(cpId) ?: throw ErrorException(ErrorType.EI_NOT_FOUND)
        if (entity.token != UUID.fromString(token)) throw ErrorException(ErrorType.INVALID_TOKEN)
        if (entity.owner != owner) throw ErrorException(ErrorType.INVALID_OWNER)
        val ei = toObject(EiDto::class.java, entity.jsonData)
        ei.apply {
            planning = ei.planning
            tender = ei.tender
        }
        entity.jsonData = toJson(ei)
        eiDao.save(entity)
        return ResponseDto(true, null, ei)
    }

    override fun getEi(cpId: String): EiDto {
        val entity = eiDao.getByCpId(cpId) ?: throw ErrorException(ErrorType.EI_NOT_FOUND)
        return toObject(EiDto::class.java, entity.jsonData)
    }

   private fun getCpId(country: String): String {
        return ocdsProperties.prefix + SEPARATOR + country + SEPARATOR + generateServise.getNowUtc()
    }

    private fun validatePeriod(ei: EiDto) {
        if (!ei.planning.budget.period.startDate.isBefore(ei.planning.budget.period.endDate))
            throw ErrorException(ErrorType.INVALID_PERIOD)
    }

    private fun getEntity(ei: EiDto, owner: String, dateTime: LocalDateTime): EiEntity {
        val ocId = ei.ocid ?: throw ErrorException(ErrorType.PARAM_ERROR)
        return EiEntity(
                cpId = ocId,
                token = generateServise.generateRandomUUID(),
                owner = owner,
                createdDate = dateTime.toDate(),
                jsonData = toJson(ei)
        )
    }

    companion object {
        private val SEPARATOR = "-"
    }

}
